Performance improvements for stratified/sorted ReadPileup

magicDGS opened the pull request:
This PR includes the following improvements for `ReadPileup` by using a new class for store `PileupElement` (`PileupElementTracker`):

* If sorting by alignment start is necessary, maintain a sorted `PileupElementTracker`
* If splitting by sample is already done, keep the stratified `PileupElementTracker`

Because pre-sorting and stratification is implemented in `LocusIteratorByState`, this will improve performance in `LocusWalker` and other tools using it. Other methods, `ReadPileup.fixOverlaps()` will benefit for these improvements.

This closes https://github.com/broadinstitute/gatk/issues/2309 and https://github.com/broadinstitute/gatk/issues/2245 (although the latest was already fixed, but it is rather inefficient).

magicDGS commented:
Hello @droazen, could you review this one? It is related with my work on the improvements for LIBS and `ReadPileup`, and it will make more efficient a lot of computation for splitting by sample the pileup object.

Thanks in advance!

magicDGS commented:
Friendly ping here, @droazen.

droazen commented:
@LeeTL1220 Since this PR possibly affects you more than anything else, would you have the time/interest to do a review pass on it?

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/SingleSamplePileupElementTracker.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
```
Can you put documentation here describing when this would be instantiated vs. the other pileup element trackers?

magicDGS replied:
Added detailed documentation everywhere.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/UnifiedPileupElementTracker.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
```
Ditto about docs, because I don't understand why the SingleSamplePileupElementTracker inherits from Unified and PerSample inherits from PileupElementTracker.  If you could just address when a developer would instantiate each.

magicDGS replied:
Added a note for the inheritance and when they are supposed to be used.

magicDGS commented:
Thanks for comments about the documentation @LeeTL1220 - I fixed them (I hope) trying to explain the logic behind the element tracker. In principle a developer shouldn't care about when to use them, because they should come from a `LocusIteratorByState` or from inside a previous tracker. In addition, the implementation should be (most of the cases) hidden from the API user, which should use `ReadPileup`.

The idea of the trackers come from GATK3, so this is a custom port with some design differences. The basic idea is to cache some operations that may be time consuming for large pileups (sorting, split by sample, extract a single sample).

I actually haven't test the performance in a proper way, just running some tools in development with the branch and it feels like is faster - in my case I use all the features that are cache: split by sample, retrieving several times single-samples and also calling `fixOverlaps()` (which uses using sorted pileups).

I think that because the `LocusIteratorByState` is already splitting by sample, that can improve even more performance, because it will come directly in the state where it can be used by-sample in an efficient way. And maybe, if the tool does not require to split by sample at all, we can add an option to disable that behavior while creating the tracker.

Looking forward for your comments and ideas about this...

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -1,61 +1,61 @@
     }
 
     /**
      * Create a new pileup at loc, using an stratified pileup
```
typo "an" --> "a"

magicDGS replied:
Done.

LeeTL1220 commented:
@magicDGS Overall, the documentation changes are good.  A few more minor comments:  
- If these are classes that are really pegged to the usage of LIBS, then you may want to mention that in the javadocs for the class.  A comment along the lines of "these classes are fairly low-level, developers should probably confirm that their changes do not belong in a higher-level calss such as LIBS".  
- I am okay with an informal test of the speed, even if you just look at some logs.  From the review, it looks like behavior of higher-level API calls will be unchanged.

>I think that because the LocusIteratorByState is already splitting by sample, that can improve even more performance, because it will come directly in the state where it can be used by-sample in an efficient way. And maybe, if the tool does not require to split by sample at all, we can add an option to disable that behavior while creating the tracker.

Agreed.  I do not think you need to worry about the flag, for now.  If you'd like, file an issue, but I think it is low priority for us.

I tend to be worried about new developers or contributors getting lost in the codebase.  And many do not have experience w/ GATK3.  Hence, the documentation about when to use the class and why it exists.

A couple additional things:
- I found a typo (see my comment)
- Can you document the `presorted` parameter?  Make sure to mention that if a developer specifies false, and sorting is needed, it will be done under-the-hood.

magicDGS commented:
@LeeTL1220 - I addressed in the next commit the documentation issues; let me know if you have more suggestions on that.

If you have any suggestion for an already implemented tool in GATK for the performance and some data for profiling, I can do a couple of runs in my local computer for look at them. If not, I can implement an small example tool for profiling when splitting by sample is needed (I can also modify `ExampleLocusWalker` for that in a different PR to being able to profile this after this PR too).

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/SingleSamplePileupElementTracker.java:
```diff
@@ -23,23 +23,47 @@
+    /**
+     * Instantiates an unsorted element tracker for the sample.
+     *
+     * <p>Note: if a {@link #sortedIterator()} is requested, sorting would be performed an cached into the tracker.
```
typo "and" --> "and"

LeeTL1220 replied:
... sorting will be performed ...

magicDGS replied:
Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/SingleSamplePileupElementTracker.java:
```diff
@@ -23,23 +23,47 @@
+    /**
+     * Instantiates a sorted/unsorted element tracker for them sample. The sample may be {@code null}.
+     *
+     * <p>Note: if {@code preSorted=true} and a {@link #sortedIterator()} is requested, sorting would be performed
```
ditto about typos

magicDGS replied:
Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/UnifiedPileupElementTracker.java:
```diff
@@ -43,12 +47,25 @@
+    /**
+     * Instantiates a sorted/unsorted element tracker.
+     *
+     * <p>Note: if {@code preSorted=true} and a {@link #sortedIterator()} is requested, sorting would be performed
```
Ditto on typos

magicDGS replied:
Done.

magicDGS commented:
Fixed typos @LeeTL1220 - maybe we can profile this with https://github.com/broadinstitute/gatk/pull/3614, which is using splitting by sample?

LeeTL1220 commented:
@magicDGS That sounds perfectly fine to use #3614 .  Regardless, I think you are safe to merge unless that quick test is worrisome.

droazen commented:
I'd like to have a quick look at this before merge, to make sure it won't negatively affect the `HaplotypeCaller`

magicDGS commented:
Waiting for your review, @droazen, for making the profiling with #3614 (maybe it would be already accepted).

magicDGS commented on src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementTrackerUnitTest.java:
```diff
@@ -0,0 +1,144 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class PileupElementTrackerUnitTest extends BaseTest {
```
Self-reminder: extend `GATKBaseTest` (requires rebase)

magicDGS commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -1,61 +1,61 @@
     public ReadPileup(final Locatable loc, final Map<String, ReadPileup> stratifiedPileup) {
-        this(loc, stratifiedPileup.values().stream().flatMap(ReadPileup::getElementStream).collect(Collectors.toList()));
+        this.loc = loc;
+        this.pileupElements = new PerSamplePileupElementTracker(stratifiedPileup.entrySet().stream()
```
Self note: this was changed in #4031 for a performance improvement for stratified pileups. We should check if there is any regression...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/utils/locusiterator/LocusIteratorByState.java:
```diff
@@ -342,7 +342,8 @@ private void lazyLoadNextAlignmentContext() {
 
                 if (!pile.isEmpty()) { // if this pileup added at least one base, add it to the full pileup
-                    fullPileupPerSample.put(sample, new ReadPileup(location, pile));
+                    // the pileup for LIBS is sorted by coordinate, because it comes from a sorted BAM file
```
This was changed by the performance improvement for HC in #4031 - we should check if maintaining the stratified pileup here cause a regression of that improvement.

droazen commented:
Re-assigning to @jonn-smith and @jamesemery to assess whether this is worth resurrecting as part of the current HaplotypeCaller performance work.

jonn-smith commented:
Closing issue due to age and code conflicts.  If this issue becomes a blocker in the future we can reinvestigate then.
